### PR TITLE
Add Scoped Custom Element Registries

### PR DESCRIPTION
**Pull request by annevk**
**Please do not comment directly on this PR unless asked**. It's a big change and we want to keep it manageable. Use #1339 instead.

HTML PR: https://github.com/whatwg/html/pull/10869.

Tests: https://github.com/web-platform-tests/wpt/pull/50790.

- [x] At least two implementers are interested (and none opposed):
   * WebKit
   * Gecko
   * Chromium
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/50790
   * https://github.com/web-platform-tests/wpt/pull/50860
   * https://github.com/web-platform-tests/wpt/pull/51953
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://issues.chromium.org/issues/399124619
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1874414
   * WebKit: N/A
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/38266
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. 

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/dom/1341.html" title="Last updated on Apr 14, 2025, 8:29 AM UTC (a77c2b8)">Preview</a> | <a href="https://whatpr.org/dom/1341/eaf2ac7...a77c2b8.html" title="Last updated on Apr 14, 2025, 8:29 AM UTC (a77c2b8)">Diff</a>

### CODE REVIEW

File: dom.bs [hunk 1]
```diff
@@ -3144,12 +3177,30 @@ otherwise, if there is no such <a for=/>element</a>, null.
...
 </pre>
 
-<p class=note>The {{DocumentOrShadowRoot}} mixin is expected to be used by other
+<dl class=domintro>
+ <dt><code><var>registry</var> = <var>documentOrShadowRoot</var> . {{DocumentOrShadowRoot/customElementRegistry}}</code>
+ <dd><p>Returns <var>documentOrShadowRoot</var>'s {{CustomElementRegistry}} object, if any;
+ otherwise null.
+</dl>
+
+<p>The <dfn attribute for=DocumentOrShadowRoot><code>customElementRegistry</code></dfn> getter steps
```
Comment (justinfagnani) [inline dom.bs:3192]: Are there setter steps for shadow roots with **keep custom element registry null** set to true?

File: dom.bs [hunk 2]
```diff
@@ -6094,12 +6232,26 @@ It is initially set to false.</p>
 <p><a for=/>Shadow roots</a> have an associated <dfn for=ShadowRoot>serializable</dfn> (a boolean).
 It is initially set to false.</p>
 
+<p><a for=/>Shadow roots</a> have an associated <dfn for=ShadowRoot>custom element registry</dfn>
+(null or a {{CustomElementRegistry}} object).</p>
+
+<p><a for=/>Shadow roots</a> have an associated
+<dfn for=ShadowRoot>keep custom element registry null</dfn> (a boolean). It is initially false.
+
+<p class=note>This can only ever be true in combination with declarative shadow roots. And it only
```
Comment (justinfagnani) [inline dom.bs:6241]: Are there any cases where we want to be able to programmatically create shadow roots that emulate declarative shadow roots with a null registry?

I think this might be the first instance where we can't imperatively make a shadow root that matches a declarative shadow root. That should be an issue for polyfilling (true polyfills after some browsers have shipped) things for DSD.

For instance, to polyfill this feature, we would need to prevent declarative shadow roots from using the global registry. So we would have to emit on the server templates with an alternate `shadowrootmode` attribute, and imperatively instantiate the shadow roots, and record the fact that it had `shadowrootcustomelementregistry` attribute.

If a future feature required a similar technique, then either we'd need to be able to set this option when creating the shadow root or be able to set a shadow root's registry to `null`.

File: dom.bs (hunk 1 above)
Comment (annevk) [reply dom.bs:3192]: No, you have to use the `initialize()` API.

File: dom.bs (hunk 2 above)
Comment (annevk) [reply dom.bs:6241]: You already can't create all declarative shadow roots. Only declarative shadow roots have the declarative bit set, for instance.

File: dom.bs [hunk 3]
```diff
@@ -6094,12 +6232,26 @@ It is initially set to false.</p>
 <p><a for=/>Shadow roots</a> have an associated <dfn for=ShadowRoot>serializable</dfn> (a boolean).
 It is initially set to false.</p>
 
+<p><a for=/>Shadow roots</a> have an associated <dfn for=ShadowRoot>custom element registry</dfn>
```
Comment (foolip) [inline dom.bs:6235]: Say that it is initially null?

**Review by foolip — COMMENTED** [review]
Note to self: editorial review done up the "Elements have an associated" bit.

File: dom.bs [hunk 4]
```diff
@@ -6499,6 +6678,26 @@ null), and boolean <var>synchronousCustomElements</var> (default false):
  <li><p>Return <var>result</var>.
 </ol>
 
+<p>To <dfn>create an element internal</dfn> given a <a for=/>document</a> <var>document</var>, an
```
Comment (foolip) [inline dom.bs:6681]: I actually am not sure what effect the `noexport` attribute has, but it's used elsewhere in DOM for spec-internal stuff, so:

```suggestion
<p>To <dfn noexport>create an element internal</dfn> given a <a for=/>document</a> <var>document</var>, an
```

Also, I first thought this might be related to `ElementInternals`, so a name like "internal create element steps" or "inner create element steps" would help avoid that. Not a big deal though.

File: dom.bs [hunk 5]
```diff
@@ -6399,74 +6574,82 @@ null), and boolean <var>synchronousCustomElements</var> (default false):
...
     <ol>
      <li><p>Let <var>C</var> be <var>definition</var>'s
      <a for="custom element definition">constructor</a>.
 
-     <li><p>Set <var>result</var> to the result of <a>constructing</a> <var>C</var>, with no
-     arguments.
-
-     <li><p>Assert: <var>result</var>'s <a for=Element>custom element state</a> and
-     <a for=Element>custom element definition</a> are initialized.
+     <li><p><a for=map>Set</a> the <a>surrounding agent</a>'s
```
Comment (foolip) [inline dom.bs:6583]: It's very hard to follow the diff here inside "create an element", so for any other reviews, I think the main things are:

- Refactoring to use a "creating an element internal" helper, which takes the registry as an argument.
- A bunch of steps nested one more level under "Run these steps while catching any exceptions", ending with the added "Set result’s custom element registry to registry."

Basically searching for "registry" within the steps shows the interesting stuff.

And I can't find anything wrong :)

File: dom.bs [hunk 6]
```diff
@@ -7118,8 +7322,8 @@ are:
 <div algorithm>
 <p>To <dfn id=concept-attach-a-shadow-root>attach a shadow root</dfn>, given an
 <a for=/>element</a> <var>element</var>, a string <var>mode</var>, a boolean <var>clonable</var>,
-a boolean <var>serializable</var>, a boolean <var>delegatesFocus</var>, and a string
-<var>slotAssignment</var>:
+a boolean <var>serializable</var>, a boolean <var>delegatesFocus</var>, a string
+<var>slotAssignment</var>, and null or a {{CustomElementRegistry}} object <var>registry</var>:
```
Comment (foolip) [inline dom.bs:7326]: If you like this style, here it is again:

```suggestion
<var>slotAssignment</var>, and a {{CustomElementRegistry}}-or-null <var>registry</var>:
```

**Review by foolip — COMMENTED** [review]
Editorial review done.

File: dom.bs (hunk 4 above)
Comment (annevk) [reply dom.bs:6681]: I mostly stopped using that as it doesn't have an effect. Except maybe to emphasize but I don't think that's needed here.

File: dom.bs [hunk 7]
```diff
@@ -4617,18 +4680,23 @@ and an optional <a for=/>document</a> <dfn export for="clone a node"><var>docume
   <p>If <var>node</var> is an <a for=/>element</a>:
 
   <ol>
+   <li><p>Let <var>registry</var> be <var>node</var>'s <a for=Element>custom element registry</a>.
```
[the remark this answers is not in the document]
Comment (annevk) [reply dom.bs:4683]: The API entry points for cloning throw for `ShadowRoot` nodes.

File: dom.bs [hunk 8]
```diff
@@ -4555,15 +4606,17 @@ and <{template}>. SVG ought to do the same for its <{script}> elements, but does
 <p>To <dfn export id=concept-node-clone>clone a node</dfn> given a <a for=/>node</a> <var>node</var>
 and an optional <a for=/>document</a> <dfn export for="clone a node"><var>document</var></dfn>
 (default <var>node</var>'s <a for=Node>node document</a>), boolean
-<dfn export for="clone a node"><var>subtree</var></dfn> (default false), and
-<a for=/>node</a>-or-null <dfn export for="clone a node"><var>parent</var></dfn> (default null):
+<dfn export for="clone a node"><var>subtree</var></dfn> (default false), <a for=/>node</a>-or-null
+<dfn export for="clone a node"><var>parent</var></dfn> (default null), and null or a
+{{CustomElementRegistry}} object <dfn export for="clone a node"><var>fallbackRegistry</var></dfn>
```
[the remark this answers is not in the document]
Comment (annevk) [reply dom.bs:4761]: This is not my preferred style. I think currently we use both on the HTML side.

[the remark this answers is not in the document]
Comment (foolip) [reply dom.bs:4763]: Alright, please disregard these suggestions.

File: dom.bs [hunk 9]
```diff
@@ -5668,6 +5756,34 @@ method steps are to return the result of running the
...
+<p>To <dfn>flatten element creation options</dfn>, given a string or {{ElementCreationOptions}}
+dictionary <var>options</var> and a <a for=/>document</a> <var>document</var>:
+
+<ol>
+ <li><p>Let <var>registry</var> be null.
+
+ <li><p>Let <var>is</var> be null.
+
+ <li>
+  <p>If <var>options</var> is a dictionary:
```
Comment (keithamus) [inline dom.bs:5768]: I think we're missing the alternate case here, for when `options` is a string - `is` is never set to `options` AFAICT.

Comment (annevk) [reply dom.bs:5768]: That is correct? It never was either.
